docs/fundamentals/networking/networking-telemetry.md

```diff
@@ -0,0 +1,547 @@
+public sealed class MyListener : EventListener
+{
+    protected override void OnEventSourceCreated(EventSource eventSource)
+    {
+        if (eventSource.Name == "System.Net.Http")
+        {
+            EnableEvents(eventSource, EventLevel.Informational);
+        }
+    }
+
+    protected override void OnEventWritten(EventWrittenEventArgs eventData)
+    {
+        Console.WriteLine($"{DateTime.UtcNow:ss:fff} {eventData.EventName}: " +
+            string.Join(' ', eventData.PayloadNames!.Zip(eventData.Payload!).Select(pair => $"{pair.First}={pair.Second}")));
+    }
+}
+```
+
+Will print something like the following:
+
+```txt
+00:598 RequestStart: scheme=https host=httpbin.org port=443 pathAndQuery=/get versionMajor=1 versionMinor=1 versionPolicy=0
+01:470 ConnectionEstablished: versionMajor=1 versionMinor=1
+01:474 RequestLeftQueue: timeOnQueueMilliseconds=470,6214 versionMajor=1 versionMinor=1
+01:476 RequestHeadersStart:
+01:477 RequestHeadersStop:
+01:592 ResponseHeadersStart:
+01:593 ResponseHeadersStop:
+01:633 ResponseContentStart:
+01:635 ResponseContentStop:
+01:635 RequestStop:
+01:637 ConnectionClosed: versionMajor=1 versionMinor=1
+```
+
+#### Yarp.Telemetry.Consumption
+
+While the `EventListener` approach outlined above is useful for quick experimentation and debugging, the APIs are not strongly-typed and force you into depending on implementation details of the instrumented library.
+
+To address this, we created a library that makes it easy to consume networking events in-process: [`Yarp.Telemetry.Consumption`].
+While the package is currently maintained as part of the [YARP] project, it can be used in any .NET application.
+
+To use it, implement the interfaces and methods (events) you are interested in:
+
+```c#
+public sealed class MyTelemetryConsumer : IHttpTelemetryConsumer, INetSecurityTelemetryConsumer
+{
+    public void OnRequestStart(DateTime timestamp, string scheme, string host, int port, string pathAndQuery, int versionMajor, int versionMinor, HttpVersionPolicy versionPolicy)
+    {
+        Console.WriteLine($"Request to {host} started at {timestamp}");
+    }
+
+    public void OnHandshakeStart(DateTime timestamp, bool isServer, string targetHost)
+    {
+        Console.WriteLine($"Starting TLS handshake with {targetHost}");
+    }
+}
+```
+
+And register the implementations with your DI container:
+
+```c#
+services.AddTelemetryConsumer<MyTelemetryConsumer>();
+```
+
+The library provides the following strongly-typed interfaces: [`IHttpTelemetryConsumer`], [`INameResolutionTelemetryConsumer`], [`INetSecurityTelemetryConsumer`], [`ISocketsTelemetryConsumer`] and [`IKestrelTelemetryConsumer`].
+
+Note that these callbacks are called as part of the instrumented operation, so the general logging guidance applies. You should avoid blocking or performing any expensive calculations as part of the callback. Offload any post-processing work to different threads to avoid adding latency to the underlying operation.
```

Comment:
```suggestion
These callbacks are called as part of the instrumented operation, so the general logging guidance applies. You should avoid blocking or performing any expensive calculations as part of the callback. Offload any post-processing work to different threads to avoid adding latency to the underlying operation.
```